### PR TITLE
stark: U-net long skip connections for wsy/wsz fine-detail injection

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Add **U-net-style long skip connections** between symmetric layers of the Transolver stack: layer i input is augmented with the output of layer (depth - 1 - i). For our 4L architecture: input of layer 3 ← output of layer 0; input of layer 2 ← output of layer 1. This is borrowed directly from modded-NanoGPT's "Skip connections" world-record entry, where adding U-net long-range skips gave a non-trivial speed/quality boost.

**Mechanism.** Standard transformer residuals add the input of each layer to its output (short-range, depth-1 skips). U-net skips add a separate path that injects shallow-layer features directly into deep-layer inputs. For CFD surrogate: the early layers carry fine-resolution geometric features (curvature, surface normals, raw position), while deep layers carry abstract aerodynamic features (slot tokens after multiple attention rounds). The wall-shear y/z gap may be partly an information-bottleneck problem: by layer 3, fine geometric detail has been compressed into the slot/slice attention bottleneck. Long skips rebroadcast the fine detail into deep layers.

**Why for wsy/wsz specifically.** wsy/wsz are dominated by sub-millimeter boundary-layer features that depend on local geometry (curvature, normal direction). A long skip that injects layer-0 features (raw post-embedding tokens) into layer-3 inputs gives the model direct access to high-resolution geometric signal at the moment it's making the final prediction.

**Risk.** Adds a 1×1 linear projection per skip (cheap). Doubles the input dimension at the merge point unless we use addition (which assumes feature alignment) or concatenation+projection. Concatenation+projection is slightly more parameters but more flexible.

## Instructions

**Code change to `target/model.py`:**

1. In the `Transformer` class (around line 205), modify the forward to support symmetric U-net long skips:
   - Cache the outputs of the first `depth // 2` layers in a list.
   - For the second half of the stack, before each layer, take its current input `x`, concatenate with the cached output of the symmetric early layer, project back to `hidden_dim` via a learned linear, and pass to the layer.
   - For 4 layers: layer 0 output cached as skip0, layer 1 output cached as skip1. Layer 2's input is `Linear(concat(x, skip1))`. Layer 3's input is `Linear(concat(x, skip0))`.

2. Add a CLI flag `--unet-skip-connections` (default `False`) that gates this behavior. When `True`, instantiate the per-skip projection modules.

3. **Run a 4-arm sweep on 4 GPUs (single-process per GPU):**

   | Arm | unet-skip | Notes |
   |---|---|---|
   | A (control) | off | Exact bengio baseline (alphonse PR #174 / m9775k1v config), no code change |
   | B | on, addition | Concatenation replaced with addition (assumes layer-i output is alignment-compatible with layer-(d-1-i) input). No new parameters. |
   | C | on, concat+project | Full version: concat then project back to hidden_dim. New params: ~2 × hidden_dim² for 4L. |
   | D | on, gated | Concat+project with a learned scalar gate `α` (initialized 0.0) so the model can opt-in to the skip information gradually. |

   All arms otherwise identical to alphonse `m9775k1v` baseline: 4L/256d/4h, AdamW, lr=3e-4, 30 epochs, cosine T_max=30, no EMA.

   **W&B group:** `stark-unet-skip-sweep`.

4. **Per-arm command (Arm B example):**

```bash
cd target/
CUDA_VISIBLE_DEVICES=1 torchrun --standalone --nproc-per-node=1 train.py \
  --agent stark --lr 3e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 8 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --epochs 30 --lr-cosine-t-max 30 --lr-warmup-steps 500 \
  --fourier-pe --fourier-pe-num-freqs 8 \
  --unet-skip-connections --unet-skip-mode addition \
  --wandb-name stark/skip-addition --wandb-group stark-unet-skip-sweep
```

## Gates

- **ep5:** kill any arm if val_abupt > 13% (alphonse ep5 was ~10.0%)
- **ep10:** kill any arm if val_abupt > 9.5% (alphonse ep10 was ~8.5%)
- **ep20:** kill any arm if val_abupt > 8.0% (alphonse ep20 was ~7.5%)
- **ep30:** beat 7.21% on any arm = win

## Win criterion

**Best skip-connection arm reaches val_abupt < 7.2091% at ep30 → merge.** Direction-of-improvement matters: if Arm D (gated) wins by a larger margin than Arms B/C, it confirms the skip information is valuable but needs slow integration. If Arm B (pure addition) matches Arm C (concat+project), the no-extra-parameter version is preferred.

**Per-axis breakdown at ep30:** Especially track wsy/wsz — if those drop disproportionately vs surf_p, that's strong evidence the information-bottleneck hypothesis is right.

## Baseline (current best on bengio)

| Metric | bengio best (val ep30) | PR | W&B run |
|---|---:|---|---|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **7.2091** | alphonse | `m9775k1v` |
| `val_primary/surface_pressure_rel_l2_pct` | **4.80** | alphonse | `m9775k1v` |
| `val_primary/wall_shear_rel_l2_pct` | (TBD) | alphonse | `m9775k1v` |

**Reproduce alphonse baseline (control arm A):**
```bash
torchrun --standalone --nproc_per_node=4 target/train.py \
  --agent alphonse --lr 3e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 8 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --epochs 30 --lr-cosine-t-max 30 --lr-warmup-steps 500 \
  --fourier-pe --fourier-pe-num-freqs 8
```

## Context

This experiment is one of the issue #252 (Modded-NanoGPT inspiration) directions — specifically item #4 in the advisor's response. PR #214 (gilbert kNN local surface attention) is currently running on bengio and showing strong wsy/wsz wins (-1.58pp wsy, -1.22pp wsz vs alphonse same-epoch at ep2) — that's evidence locality matters for wall-shear; long skips are a different mechanism that achieves the same goal (more local detail in the deep layers). The two experiments are complementary, not duplicative.

## Implementation notes

- The `_apply_token_mask` calls in TransformerBlock.forward should still be applied at the start of each block. Apply the skip injection BEFORE the mask, so masked tokens are zeroed in both the main and skip paths.
- For the gated variant (Arm D), initialize `α = 0` so initial behavior matches the control. Use `nn.Parameter(torch.zeros(1))` and apply as `x = x + α * skip_proj`.
- Verify the skip projection layers are in the correct param group for your optimizer (AdamW gets all params here, no Muon split needed).

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #266 because ADVISOR: Closing this PR — the branch name `stark/post-attn-rmsnorm` (post-attention RMSNorm) does not match any assigned Wave 5 hypothesis for student stark, and there is no `senpai-bengio-stark` pod provisioned in the cluster. No experiment was run and no results are available to review. If you believe this was an error in assignment, please reach out and a new hypothesis will be assigned on a fresh branch. GPU slots are too valuable to hold open on an unprovisioned experiment..
